sdk/metric/exemplar/fixed_size_reservoir.go

```diff
@@ -127,25 +101,58 @@ func (r *FixedSizeReservoir) Offer(ctx context.Context, t time.Time, n Value, a
 	// https://github.com/MrAlias/reservoir-sampling for a performance
 	// comparison of reservoir sampling algorithms.
 
-	r.mu.Lock()
-	defer r.mu.Unlock()
-	if int(r.count) < cap(r.measurements) {
-		r.store(ctx, int(r.count), t, n, a)
-	} else if r.count == r.next {
+	count, next := r.incrementCount()
+	intCount := int(count) // nolint:gosec // count is at most 32 bits in length
+	if intCount < r.k {
+		r.store(ctx, intCount, t, n, a)
+	} else if count == next {
 		// Overwrite a random existing measurement with the one offered.
-		idx := int(rand.Int64N(int64(cap(r.measurements))))
+		idx := rand.IntN(r.k)
 		r.store(ctx, idx, t, n, a)
+		r.wMu.Lock()
+		defer r.wMu.Unlock()
 		r.advance()
 	}
-	r.count++
+}
+
+// Collect returns all the held exemplars.
+//
+// The Reservoir state is preserved after this call.
+func (r *FixedSizeReservoir) Collect(dest *[]Exemplar) {
+	r.storage.Collect(dest)
+	// Call reset here even though it will reset r.count and restart the random
+	// number series. This will persist any old exemplars as long as no new
+	// measurements are offered, but it will also prioritize those new
+	// measurements that are made over the older collection cycle ones.
+	r.reset()
+}
+
+func newNextTracker(k int) *nextTracker {
+	nt := &nextTracker{k: k}
+	nt.reset()
+	return nt
+}
+
+type nextTracker struct {
+	// count is the number of measurement seen, and is in the lower 32 bits.
+	// once the reservoir has been filled, and is in the upper 32 bits.
+	countAndNext atomic.Uint64
+	// w is the largest random number in a distribution that is used to compute
+	// the next next.
+	w float64
+	// wMu ensures w is kept consistent with next during advance and reset.
+	wMu sync.Mutex
+	// k is the number of measurements that can be stored in the reservoir.
+	k int
```

Comment:
Shouldn't `k` be `uint32` here given that is the max size now supported?

```suggestion
	k uint32
```